### PR TITLE
Drop old C++ standards

### PR DESCRIPTION
**Pull request by jspricke**

### CODE REVIEW

**Review by rhaschke — COMMENTED** [review]
As you see from CI results, these definitions are required for Bionic.
Why, in first place, do you want to get rid of those standard definitions?

**Comment by jspricke** [comment]
>As you see from CI results, these definitions are required for Bionic.

But the failing test is in Focal (which should have C++14) and according 
to Henning master only targets noetic. I will try to reproduce it later.

>Why, in first place, do you want to get rid of those standard definitions?

Debian (unstable) got the new log4cxx which depends on C++17 and is used 
by rosconsole. These flags force the catkin_ws to an older version and 
thus fail to compile. See it as a preparation for Ubuntu 22.04 ;).

**Comment by rhaschke** [comment]
> But the failing test is in Focal (which should have C++14)
No, the failing test is in Bionic: the docker image `moveit/moveit:master-ci` is based on Melodic.

> according to @henningkayser master only targets noetic. I will try to reproduce it later.

No, the `master` branch also targets Melodic. As far as I remember we even recommend building the master branch on Melodic to get the latest stuff there.
So we do need an approach to require C++14 _at least_.

**Comment by jspricke** [comment]
>No, the failing test is in Bionic: the docker image `moveit/moveit:master-ci` is based on Melodic.

Ah, thanks!

>No, the `master` branch also targets Melodic. As far as I remember we even recommend building the master branch on Melodic to get the latest stuff there.

So why is there a melodic-devel branch pointing to a different state 
then master?

>So we do need an approach to require C++14 _at least_.

That would be possible with target_compile_features but that seems like 
a lot of effort. I guess we can just wait till you drop the Bionic 
requirement from master. Feel free to close this until then.

**Comment by henningkayser** [comment]
huh, I thought melodic would have been dropped by now, hence the noetic-devel branches in some of the upstream dependencies https://github.com/ros-planning/moveit/blob/master/moveit.rosinstall. Sorry @jspricke for the confusion.

**Comment by rhaschke** [comment]
Of course, master _primarily_ targets Noetic. However, it _also_ targets Melodic systems.

**Comment by rhaschke** [comment]
I don't quite understand why having different (minimal) C++ standard definitions in a project and its upstream dependencies doesn't work: I would expect cmake to pick the _highest_ required standard.

**Comment by jspricke** [comment]
>I don't quite understand why having different (minimal) C++ standard definitions in a project and its upstream dependencies doesn't work: I would expect cmake to pick the _highest_ required standard.

To my understanding that behavior is implemented in 
target_compile_features whereas set(CMAKE_CXX_STANDARD 14) and 
add_definitions(-std=c++14) force the compiler to the old standard.

**Comment by gavanderhoorn** [comment]
> I don't quite understand why having different (minimal) C++ standard definitions in a project and its upstream dependencies doesn't work: I would expect cmake to pick the _highest_ required standard.

if those packages export that requirement, and the dependent packages also consume that information, then yes, that should work.

But that's "modern CMake", and Catkin doesn't do that automatically for you.

**Comment by jspricke** [comment]
I think the current version should be fine for all cases.

**Review by rhaschke — COMMENTED** [review]
Looks good for gcc. What about clang?

**Comment by jspricke** [comment]
>Looks good for gcc. What about clang?

Clang 3.4 and higher uses 14 by default and Bionic has 6, so I assume 
that is safe.

>> @@ -1,9 +1,6 @@
> cmake_minimum_required(VERSION 3.1.3)
> project(pilz_industrial_motion_planner)
>
>-## Add support for C++14, supported in ROS Melodic and newer
>-add_definitions(-std=c++14)
>
>I think this should be replaced with `moveit_build_options()`.

Why would that be needed, shouldn't it get the definition from 
moveit_core?

>> @@ -1,9 +1,6 @@
> cmake_minimum_required(VERSION 3.1.3)
> project(pilz_industrial_motion_planner_testutils)
>
>-## Add support for C++11, supported in ROS Kinetic and newer
>-add_definitions(-std=c++11)
>
>Same here.

Why would you need 11, that's the default even on Bionic.

File: moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt [hunk 1]
```diff
@@ -1,9 +1,6 @@
 cmake_minimum_required(VERSION 3.1.3)
 project(pilz_industrial_motion_planner)
 
-## Add support for C++14, supported in ROS Melodic and newer
-add_definitions(-std=c++14)
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt:5]: True.

**Pull request merged**